lib/rmg/CMapGenerator.cpp

```diff
@@ -268,34 +270,66 @@ void CMapGenerator::addPlayerInfo()
 		map->players[pSettings.getColor().getNum()] = player;
 	}
 
-	map->howManyTeams = (mapGenOptions->getTeamCount() == 0 ? mapGenOptions->getPlayerCount() : mapGenOptions->getTeamCount())
-			+ (mapGenOptions->getCompOnlyTeamCount() == 0 ? mapGenOptions->getCompOnlyPlayerCount() : mapGenOptions->getCompOnlyTeamCount());
+	map->howManyTeams = (mapGenOptions.getTeamCount() == 0 ? mapGenOptions.getPlayerCount() : mapGenOptions.getTeamCount())
+			+ (mapGenOptions.getCompOnlyTeamCount() == 0 ? mapGenOptions.getCompOnlyPlayerCount() : mapGenOptions.getCompOnlyTeamCount());
 }
 
 void CMapGenerator::genZones()
 {
-	editManager->clearTerrain(&rand);
-	editManager->getTerrainSelection().selectRange(MapRect(int3(0, 0, 0), mapGenOptions->getWidth(), mapGenOptions->getHeight()));
-	editManager->drawTerrain(ETerrainType::GRASS, &rand);
+	getEditManager()->clearTerrain(&rand);
+	getEditManager()->getTerrainSelection().selectRange(MapRect(int3(0, 0, 0), mapGenOptions.getWidth(), mapGenOptions.getHeight()));
+	getEditManager()->drawTerrain(ETerrainType::GRASS, &rand);
+	
 
-	auto tmpl = mapGenOptions->getMapTemplate();
+	auto tmpl = mapGenOptions.getMapTemplate();
 	zones.clear();
 	for(const auto & option : tmpl->getZones())
 	{
-		auto zone = std::make_shared<CRmgTemplateZone>();
-		zone->setOptions(option.second.get());
+		auto zone = std::make_shared<CRmgTemplateZone>(this);
+		zone->setOptions(*option.second.get());
 		zones[zone->getId()] = zone;
-		//todo: move to CRmgTemplateZone constructor
-		zone->setGenPtr(this);//immediately set gen pointer before taking any actions on zones
 	}
 
 	CZonePlacer placer(this);
-	placer.placeZones(mapGenOptions, &rand);
-	placer.assignZones(mapGenOptions);
+	placer.placeZones(&rand);
+	placer.assignZones();
+	
+	//add special zone for water
+	zoneWater.first = zones.size()+1;
+	zoneWater.second = std::make_shared<CRmgTemplateZone>(this);
+	{
+		rmg::ZoneOptions options;
+		options.setId(zoneWater.first);
+		options.setType(ETemplateZoneType::WATER);
+		zoneWater.second->setOptions(options);
+	}
 
 	logGlobal->info("Zones generated successfully");
 }
 
+void CMapGenerator::createWaterTreasures()
+{
+	ui32 waterTreasure = getZoneWater().second->getTileInfo().size();
+	//add treasures on water
+	getZoneWater().second->addTreasureInfo(CTreasureInfo{1000, 3000, 9});
+	getZoneWater().second->addTreasureInfo(CTreasureInfo{5000, 10000, 1});
+	if(waterTreasure > 1000)
+		getZoneWater().second->addTreasureInfo(CTreasureInfo{2000, 6000, 2});
+	if(waterTreasure > 2500)
+		getZoneWater().second->addTreasureInfo(CTreasureInfo{2000, 6000, 2});
+	if(waterTreasure > 3600)
+		getZoneWater().second->addTreasureInfo(CTreasureInfo{2000, 6000, 3});
+	if(waterTreasure > 6400)
+		getZoneWater().second->addTreasureInfo(CTreasureInfo{2000, waterTreasure, 4});
```

Comment:
There's no need to count area, just define treasure **density** and it will be ok.

You can also ask HoTA team for exact values, I can't find them anymore :(

Reply:
Hey, got it: https://forum.df2.ru/index.php?showtopic=32286

> Changes to the treasures of the water zone.
> How it was.
> Value 100-1000: Frequency of occurence 5.
> Value 2000-6000: Frequency of occurence 1.
> 
> How it is now.
> Value 100-1000: Frequency of occurence 6.
> Value 1000-4000: Frequency of occurence 3.
> Value 4000-12000: Frequency of occurence 1.

Reply:
Keep in mind though that there are less water objects to choose from in vanilla.

Reply:
I've decided to use initial variant, but in some other PR will move it to json